File: earthsciio/backends/file.py
"""The active ``file`` transport — a local copy into the cache.

Registered under scheme ``file`` (``spec/registries.md`` §1). It resolves a
``file://`` URL to a local path — expanding ``${EARTHSCIDATADIR}`` first (the
``nei2016`` mirror pattern, ``spec/cache-format.md`` §5) — and copies it into the
staging path. ``file://`` has no conditional-GET or auth semantics, so those
arguments are accepted (for interface conformance) and ignored; the result is
always :data:`~earthsciio.transport.DOWNLOADED` with no validators.

This is what lets a pre-populated local mirror — or the conformance corpus —
feed the cache with exactly the same code path as a network fetch.
"""

from __future__ import annotations

import os
import shutil
from typing import Any, Dict, List, Optional

from ..errors import TransportError
from ..transport import DOWNLOADED, FetchResult, file_url_to_path

__all__ = ["FileTransport"]


class FileTransport:
    """Copy a ``file://`` source (or bare path) into the staging file."""

    NAME = "file"
    SCHEMES = ("file",)

    def schemes(self) -> List[str]:
        return list(self.SCHEMES)

    def fetch(
        self,
        resolved_url: str,
        dest: Any,
        conditional: Optional[Dict[str, Any]] = None,
        auth: Optional[Any] = None,
    ) -> FetchResult:
        src = file_url_to_path(resolved_url)
        if not os.path.isfile(src):
            raise TransportError(
                f"file transport: source not found: {src} (from {resolved_url})"
            )
        try:
            shutil.copyfile(src, os.fspath(dest))
        except OSError as exc:
            raise TransportError(
                f"file copy failed {src} -> {dest}: {exc}"
            ) from exc
        return FetchResult(
            DOWNLOADED,
            etag=None,
            last_modified=None,
            bytes_written=os.path.getsize(dest),
        )

File: rust/src/transport/file.rs
//! The `file` transport: copies a local file into the cache
//! (`spec/registries.md` §1). Expands `${EARTHSCIDATADIR}` in the path so a
//! pre-populated local mirror is found (`spec/cache-format.md` §5). No
//! conditional GET / auth — local files have no HTTP validators.

use std::path::{Path, PathBuf};

use super::{Conditional, FetchResult, FetchStatus, Transport};
use crate::auth::AuthResolver;
use crate::datadir::expand_datadir;
use crate::error::{Error, Result};

/// `file://` transport.
pub struct FileTransport;

impl FileTransport {
    /// Construct the file transport.
    pub fn new() -> Self {
        Self
    }
}

impl Default for FileTransport {
    fn default() -> Self {
        Self::new()
    }
}

impl Transport for FileTransport {
    fn schemes(&self) -> &'static [&'static str] {
        &["file"]
    }

    fn fetch(
        &self,
        url: &str,
        dest: &Path,
        _conditional: &Conditional,
        _auth: Option<&dyn AuthResolver>,
    ) -> Result<FetchResult> {
        let src = file_url_to_path(url)?;
        let bytes_written =
            std::fs::copy(&src, dest).map_err(|e| Error::io(Some(src.clone()), e))?;
        Ok(FetchResult {
            status: FetchStatus::Downloaded,
            etag: None,
            last_modified: None,
            bytes_written,
        })
    }
}

/// Turn a `file://` URL into a local path, expanding `${EARTHSCIDATADIR}` first.
///
/// Handles `file:///abs/path` (empty authority) and `file://host/abs/path`
/// (authority dropped). The path is taken literally after expansion — no
/// percent-decoding (the resolved URLs the cache produces are not encoded).
fn file_url_to_path(url: &str) -> Result<PathBuf> {
    let expanded = expand_datadir(url);
    let rest = expanded
        .strip_prefix("file://")
        .ok_or_else(|| Error::BadUrl {
            url: url.to_string(),
            detail: "not a file:// URL".to_string(),
        })?;

    let path = if let Some(stripped) = rest.strip_prefix('/') {
        // `file:///abs` ⇒ rest = "/abs" ⇒ keep the leading slash.
        format!("/{stripped}")
    } else {
        // `file://host/abs` ⇒ drop the authority, path starts at the first '/'.
        match rest.find('/') {
            Some(i) => rest[i..].to_string(),
            None => {
                return Err(Error::BadUrl {
                    url: url.to_string(),
                    detail: "file:// URL has no path".to_string(),
                })
            }
        }
    };
    Ok(PathBuf::from(path))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_triple_slash_abs_path() {
        assert_eq!(
            file_url_to_path("file:///scratch/x.nc").unwrap(),
            PathBuf::from("/scratch/x.nc")
        );
    }

    #[test]
    fn drops_authority() {
        assert_eq!(
            file_url_to_path("file://localhost/scratch/x.nc").unwrap(),
            PathBuf::from("/scratch/x.nc")
        );
    }

    #[test]
    fn rejects_non_file_url() {
        assert!(file_url_to_path("https://x/y").is_err());
    }
}
